zk.go

```diff
@@ -176,3 +189,96 @@ func LeaderFromZKOpts(options ...ZKOpt) (string, error) {
 
 	return leaderURL, nil
 }
+
+// Retrieves current mesos leader from ZK with a custom configuration.
+func MesosFromZKOpts(options ...ZKOpt) (string, error) {
+	var mesosURL string
+
+	// Load the default configuration for Zookeeper followed by overriding values with those provided by the caller.
+	config := &zkConfig{backoff: defaultBackoff, timeout: time.Second * 10, logger: NoopLogger{}}
+	for _, opt := range options {
+		opt(config)
+	}
+
+	if len(config.endpoints) == 0 {
+		return "", errors.New("no Zookeeper endpoints supplied")
+	}
+
+	if config.path == "" {
+		return "", errors.New("no Zookeeper path supplied")
+	}
+
+	// Create a closure that allows us to use the ExponentialBackoff function.
+	config.backoff.Steps = 1
+	retryErr := ExponentialBackoff(config.backoff, config.logger, func() (bool, error) {
+
+		c, _, err := zk.Connect(config.endpoints, config.timeout, func(c *zk.Conn) { c.SetLogger(config.logger) })
+		if err != nil {
+			return false, NewTemporaryError(errors.Wrap(err, "Failed to connect to Zookeeper"))
+		}
+
+		defer c.Close()
+
+		// Open up descriptor for the ZK path given
+		children, _, _, err := c.ChildrenW(config.path)
+		if err != nil {
+
+			// Sentinel error check as there is no other way to check.
+			if err == zk.ErrInvalidPath {
+				return false, errors.Wrapf(err, "path %s is an invalid Zookeeper path", config.path)
+			}
+
+			return false,
+				NewTemporaryError(errors.Wrapf(err, "path %s doesn't exist on Zookeeper ", config.path))
+		}
+
+		// Search for the leader through all the children in the given path
+		minScore := math.MaxInt64
+		mesosInstance := new(MesosInstance)
+		for _, child := range children {
+
+			// Only the leader will start with json.info_
+			if strings.HasPrefix(child, "json.info_") {
+				score, err := strconv.Atoi((strings.Split(child, "_"))[1])
```

Comment:
Possible out of bounds error here if string is malformed. We should check that the len of the split is 2.